### PR TITLE
[JIT Kernel][1/2]Migrate MXFP8 Group GEMM & Quant into JIT

### PR DESCRIPTION
**Pull request by HydraQYH**
## Motivation
Migrate MXFP8 Group GEMM & Quant into JIT. **Note that this is not a simple transplant.** I used MoE Group GEMM instead of the normal Group GEMM([Reference CUTLASS Example](https://github.com/NVIDIA/cutlass/blob/main/examples/92_blackwell_moe_gemm/92_blackwell_moe_gemm_blockscaled_rcgrouped.cu)). This implementation reduces TMA descriptor updates. In addition, we used a 2SM strategy to reduce data copying and improve performance in certain memory-bound scenarios.

## Modifications
Just look at the changes in the file.

## Accuracy Tests
The unit tests included in the PR have passed.

## Speed Tests and Profiling
```
mxfp8-moe-groupmm-performance:
    total_tokens     n_g     k_g  num_experts  JIT MXFP8 MoE GroupMM  sgl-kernel MXFP8 MoE GroupMM
0         1024.0  4096.0  4096.0         64.0             180.088000                    208.037888
1         2048.0  4096.0  4096.0         64.0             190.629537                    211.969478
2         4096.0  4096.0  4096.0         64.0             190.435118                    222.083860
3           32.0  4096.0  7168.0         32.0             156.727480                    178.865933
4           32.0  7168.0  2048.0         32.0              86.967122                    105.236288
5           64.0  4096.0  7168.0         32.0             155.089191                    182.351810
6           64.0  7168.0  2048.0         32.0              86.726145                    104.316447
7          128.0  4096.0  7168.0         32.0             155.794210                    182.123855
8          128.0  7168.0  2048.0         32.0              87.850831                    103.912049
9          256.0  4096.0  7168.0         32.0             158.440509                    183.737484
10         256.0  7168.0  2048.0         32.0              88.154868                    104.278600
11         512.0  4096.0  7168.0         32.0             159.324559                    183.736998
12         512.0  7168.0  2048.0         32.0              90.343090                    106.248101
13        1024.0  4096.0  7168.0         32.0             166.528880                    187.202076
14        1024.0  7168.0  2048.0         32.0              91.561101                    106.099247
15        2048.0  4096.0  7168.0         32.0             169.180446                    192.767996
16        2048.0  7168.0  2048.0         32.0              97.343141                    107.879057
17        4096.0  4096.0  7168.0         32.0             177.959908                    200.241027
18        4096.0  7168.0  2048.0         32.0             103.720810                    114.993133
19        8192.0  4096.0  7168.0         32.0             300.926693                    313.530947
20        8192.0  7168.0  2048.0         32.0             160.460002                    190.991465
```
Let's take `m_per_group=128, n=4096, k=4096, groups=64` as an example to illustrate the reasons for the performance improvement. First, fewer TMA descriptor updates make TensorCore utilization smoother:
Normal Group GEMM:
image
MoE Group GEMM:
image
Secondly, the 2SM strategy can reduce the number of data copies, save shared memory usage, and effectively improve performance in memory-bound scenarios:
Normal Group GEMM:
image
MoE Group GEMM:
image
Data transfer volume calculation:
PR Image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/jit_kernel/benchmark/bench_mxfp8_moe.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        (a_blockscale_offset, k_g // 32), dtype=torch.uint8, device=device
+    )
+
+    b_quant = torch.zeros_like(b, dtype=torch.float8_e4m3fn, device=device)
+    b_scale_factor = torch.zeros(
+        (num_experts * n_g, k_g // 32), dtype=torch.uint8, device=device
+    )
+
+    # Use a global workspace to avoid allocating 1GB every time
+    workspace = torch.empty((1024, 1024, 1024), dtype=torch.uint8, device=device)
```
[the remark this answers is not in the document]
Comment (HydraQYH) [reply python/sglang/jit_kernel/benchmark/bench_mxfp8_moe.py:129]: Not right.

File: python/sglang/jit_kernel/csrc/moe/expert_specialization/es_sm100_mxfp8_blockscaled_moe_group_gemm.cuh [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+      reinterpret_cast<int*>(expert_offsets.data_ptr()),
+      reinterpret_cast<int*>(blockscale_offsets.data_ptr()),
+      reinterpret_cast<ElementB*>(b.data_ptr()),
+      reinterpret_cast<ElementSF*>(sfb.data_ptr()),
+      reinterpret_cast<ElementD*>(d.data_ptr()),
+      reinterpret_cast<ElementB**>(b_ptrs.data_ptr()),
+      reinterpret_cast<ElementSF**>(sfb_ptrs.data_ptr()),
+      reinterpret_cast<ElementD**>(d_ptrs.data_ptr()));
+
+  sm100_mxfp8_blockscaled_moe_group_gemm_pre_compute_kernel<<<1, num_experts, 0, stream>>>(offset_functor, m, k);
```
[the remark this answers is not in the document]
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/moe/expert_specialization/es_sm100_mxfp8_blockscaled_moe_group_gemm.cuh:46]: Not needed.

File: python/sglang/jit_kernel/csrc/moe/expert_specialization/es_sm100_mxfp8_blockscaled_group_quant.cuh [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+  cuda::ptx::fence_proxy_async(cuda::ptx::space_shared);  // b)
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    cuda::ptx::cp_async_bulk(
+        cuda::ptx::space_global,
+        cuda::ptx::space_shared,
+        squeeze_tiled_tensor_sf.data().get(),
+        squeeze_tiled_tensor_shared_sf.data().get(),
+        512);
```
[the remark this answers is not in the document]
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/moe/expert_specialization/es_sm100_mxfp8_blockscaled_group_quant.cuh:256]: Not needed.

**Review by BBuf — APPROVED** [review]
